realm/realm-library/src/androidTest/java/io/realm/RealmCacheTests.java

```diff
@@ -331,6 +335,181 @@ public void releaseCacheInOneThread() {
         assertNull(realmA.sharedRealm);
     }
 
+    @Test
+    @RunTestInLooperThread
+    public void getInstanceAsync_typedRealm() {
+        final RealmConfiguration configuration = looperThread.createConfiguration();
+        final AtomicBoolean realmCreated = new AtomicBoolean(false);
+        Realm.getInstanceAsync(configuration, new RealmInstanceCallback<Realm>() {
+            @Override
+            public void onSuccess(Realm realm) {
+                realmCreated.set(true);
+                assertEquals(1, Realm.getLocalInstanceCount(configuration));
+                realm.close();
+                looperThread.testComplete();
+            }
+        });
+        assertFalse(realmCreated.get());
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void getInstanceAsync_dynamicRealm() {
+        final RealmConfiguration configuration = looperThread.createConfiguration();
+        final AtomicBoolean realmCreated = new AtomicBoolean(false);
+        DynamicRealm.getInstanceAsync(configuration, new RealmInstanceCallback<DynamicRealm>() {
+            @Override
+            public void onSuccess(DynamicRealm realm) {
+                realmCreated.set(true);
+                assertEquals(1, Realm.getLocalInstanceCount(configuration));
+                realm.close();
+                looperThread.testComplete();
+            }
+        });
+        assertFalse(realmCreated.get());
+    }
+
+    @Test
+    @RunTestInLooperThread
+    public void getInstanceAsync_callbackImmediatelyWhenLocalCacheExist() {
```

Comment:
IMO I don't like this semantic. As discussed in https://github.com/realm/realm-java/issues/4270#issuecomment-285330690 

While technically it isn't wrong per the contract of async to do that, it can and probably will mess up people's mental model. 

I would much rather have all callbacks happen through the looper queue, which will allow all remaining setup code in e.g. `onCreate` to run before this is called.